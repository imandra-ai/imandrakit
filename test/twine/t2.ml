type x1 =
  | A of int
  | B
  | C of {
      x: int option;
      z: (bool * char) list;
    }
  | D of x1 * x1
[@@deriving twine, show { with_path = false }] [@@hashcons]

type x1_alias = x1 [@@deriving twine, show]

type x2 = {
  a: x1_alias;
  b: x1 list list array;
  c: x2 list option;
}
[@@deriving twine, show { with_path = false }]
[@@twine.use_field_names]
[@@hashcons]

let c0 = C { x = Some 3; z = [ true, 'a'; false, 'c' ] }

let myx2 =
  let c0 = C { x = Some 3; z = [ true, 'a'; false, 'c' ] } in
  let c1 = B in
  {
    a = A 42;
    b = [| [ []; [ c0 ] ]; [ [ B; c1; c1 ]; [ c0 ]; [ c0; D (c0, c0) ] ] |];
    c = None;
  }

let myx2_big = { a = B; b = [| [ []; [ c0; c0 ] ] |]; c = Some [ myx2; myx2 ] }
;;

Format.printf "start with myx2:@.%a@." pp_x2 myx2

let s = Imandrakit_twine.Encode.to_string x2_to_twine myx2
let () = Printf.printf "serialized to %S\n" s
let () = Printf.printf "len: %d\n" (String.length s)
let s' = Marshal.to_string myx2 [ Marshal.No_sharing ]
let () = Printf.printf "len with marshal: %d\n" (String.length s')
let () = Format.printf "dump:@.%s@." (Imandrakit_twine.Dump.dump_string s)
let foo2 = Imandrakit_twine.Decode.decode_string x2_of_twine s
let () = Format.printf "myx2 after roundtrip:@.%a@." pp_x2 foo2;;

assert (foo2 = myx2)

let s_big = Imandrakit_twine.Encode.to_string x2_to_twine myx2_big
let () = Printf.printf "len: %d\n" (String.length s_big)
let s'_big = Marshal.to_string myx2_big [ Marshal.No_sharing ]
let () = Printf.printf "len(big) with marshal: %d\n" (String.length s'_big)
let foo2_big = Imandrakit_twine.Decode.decode_string x2_of_twine s_big
let () = Format.printf "deser(myx2_big):@.%a@." pp_x2 foo2_big

let () =
  Format.printf "myx2_big hex is:@.%s@." (Hex.hexdump_s @@ Hex.of_string s_big)
;;

assert (myx2_big = foo2_big)
