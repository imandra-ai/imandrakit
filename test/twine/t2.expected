start with myx2:
{ a = (A 42);
  b =
  [|[[]; [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}]];
    [[B; B; B]; [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}];
      [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]};
        (D (C {x = (Some 3); z = [(true, 'a'); (false, 'c')]},
           C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}))
        ]
      ]
    |];
  c = None }
serialized to "`b\001\031Rb\000\031Tb\248\245\019a\241\194\002\243\248a\244b\255\006\244\161\161\161c\243\243\243b\001\031Rb\000\031Tb\248\245\019a\241\194\002\243\248a\244b\001\031Rb\000\031Tb\248\245\019a\241\194\002\243\248b\001\031Rb\000\031Tb\248\245\019a\241\194\002\243\248b\001\031Rb\000\031Tb\248\245\019a\241\194\002\243\248\195\002\245\255\tb\255\030\247c\255H\2554\248b\255U\248\176\031\027c\243\248\002\003"
len: 133
len with marshal: 118
dump:
Array(3) [
 C_0[42],
 Array(2) [
   Array(2) [
     [] ],
     Array(1) [
       C_2[[3],[[true,97],[false,99]]] ] ],
   Array(3) [
     Array(3) [
       C_1[],
       C_1[],
       C_1[] ],
     Array(1) [
       C_2[[3],[[true,97],[false,99]]] ],
     Array(2) [
       C_2[[3],[[true,97],[false,99]]],
       C_3[C_2[[3],[[true,97],[false,99]]],C_2[[3],[[true,97],[false,99]]]] ] ] ],
 null ]
myx2 after roundtrip:
{ a = (A 42);
  b =
  [|[[]; [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}]];
    [[B; B; B]; [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}];
      [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]};
        (D (C {x = (Some 3); z = [(true, 'a'); (false, 'c')]},
           C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}))
        ]
      ]
    |];
  c = None }
len: 324
len(big) with marshal: 258
deser(myx2_big):
{ a = B;
  b =
  [|[[];
      [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]};
        C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}]
      ]
    |];
  c =
  (Some [{ a = (A 42);
           b =
           [|[[]; [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}]];
             [[B; B; B]; [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}];
               [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]};
                 (D (C {x = (Some 3); z = [(true, 'a'); (false, 'c')]},
                    C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}))
                 ]
               ]
             |];
           c = None };
          { a = (A 42);
            b =
            [|[[]; [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}]];
              [[B; B; B];
                [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}];
                [C {x = (Some 3); z = [(true, 'a'); (false, 'c')]};
                  (D (C {x = (Some 3); z = [(true, 'a'); (false, 'c')]},
                     C {x = (Some 3); z = [(true, 'a'); (false, 'c')]}))
                  ]
                ]
              |];
            c = None }
          ])
  }
myx2_big hex is:
00000000: 6062 011f 5262 001f 5462 f8f5 1361 f1c2  `b..Rb..Tb...a..
00000001: 02f3 f861 f462 ff06 f4a1 a1a1 63f3 f3f3  ...a.b......c...
00000002: 6201 1f52 6200 1f54 62f8 f513 61f1 c202  b..Rb..Tb...a...
00000003: f3f8 61f4 6201 1f52 6200 1f54 62f8 f513  ..a.b..Rb..Tb...
00000004: 61f1 c202 f3f8 6201 1f52 6200 1f54 62f8  a.....b..Rb..Tb.
00000005: f513 61f1 c202 f3f8 6201 1f52 6200 1f54  ..a.....b..Rb..T
00000006: 62f8 f513 61f1 c202 f3f8 c302 f5ff 0962  b...a..........b
00000007: ff1e f763 ff48 ff34 f862 ff55 f8b0 1f1b  ...c.H.4.b.U....
00000008: 63f3 f802 6062 011f 5262 001f 5462 f8f5  c...`b..Rb..Tb..
00000009: 1361 f1c2 02f3 f861 f462 ff06 f4a1 a1a1  .a.....a.b......
00000010: 63f3 f3f3 6201 1f52 6200 1f54 62f8 f513  c...b..Rb..Tb...
00000011: 61f1 c202 f3f8 61f4 6201 1f52 6200 1f54  a.....a.b..Rb..T
00000012: 62f8 f513 61f1 c202 f3f8 6201 1f52 6200  b...a.....b..Rb.
00000013: 1f54 62f8 f513 61f1 c202 f3f8 6201 1f52  .Tb...a.....b..R
00000014: 6200 1f54 62f8 f513 61f1 c202 f3f8 c302  b..Tb...a.......
00000015: f5ff 0962 ff1e f763 ff48 ff34 f862 ff55  ...b...c.H.4.b.U
00000016: f8b0 1f1b 63f3 f802 62ff 79f6 61f4 6062  ....c...b.y.a.`b
00000017: 011f 5262 001f 5462 f8f5 1361 f1c2 02f3  ..Rb..Tb...a....
00000018: f862 011f 5262 001f 5462 f8f5 1361 f1c2  .b..Rb..Tb...a..
00000019: 02f3 f862 ff07 f662 ff1a f661 f4a1 63f1  ...b...b...a..c.
00000020: f4ff 2504                                ..%.

